tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,378 @@
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  TestParam test_case_with_different_dtypes = {
+      /*input_tensors*/ {
+          {CreateTensor<int64>(TensorShape({2, 2}), {1, 2, 3, 4})},
+          {CreateTensor<double>(TensorShape({2, 2}), {1.0, 2.0, 3.0, 4.0})}},
+      /*expected_outputs*/ {},
+      /*expected_output_dtypes*/ {DT_INT64},
+      /*expected_output_shapes*/ {PartialTensorShape({2})},
+      /*expected_cardinality*/ 0,
+      /*breakpoints*/ {}};
+
+  EXPECT_EQ(CreateDatasetFromTestCase(test_case_with_different_dtypes).code(),
+            tensorflow::error::INVALID_ARGUMENT);
+}
+
+TEST_F(ConcatenateDatasetOpTestHelper, DatasetName) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+  TF_ASSERT_OK(CreateDatasetFromTestCase(TestCases[0]));
+
+  EXPECT_EQ(dataset->name(), kOpName);
+}
+
+struct DatasetOutputDtypesTest : ConcatenateDatasetOpTestHelper,
+                                 ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetOutputDtypesTest, DatasetOutputDtypes) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+  const TestParam &test_case = GetParam();
+  TF_ASSERT_OK(CreateDatasetFromTestCase(test_case));
+  EXPECT_OK(VerifyTypesMatch(dataset->output_dtypes(),
+                             test_case.expected_output_dtypes));
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetOutputDtypesTest,
+                        ::testing::ValuesIn(TestCases));
+
+struct DatasetOutputShapesTest : ConcatenateDatasetOpTestHelper,
+                                 ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetOutputShapesTest, DatasetOutputShapes) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+  const TestParam &test_case = GetParam();
+  TF_ASSERT_OK(CreateDatasetFromTestCase(test_case));
+  EXPECT_OK(VerifyShapesCompatible(dataset->output_shapes(),
+                                   test_case.expected_output_shapes));
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetOutputShapesTest,
+                        ::testing::ValuesIn(TestCases));
+
+struct DatasetCardinalityTest : ConcatenateDatasetOpTestHelper,
+                                ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetCardinalityTest, Cardinality) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+  const TestParam &test_case = GetParam();
+  TF_ASSERT_OK(CreateDatasetFromTestCase(test_case));
+
+  EXPECT_EQ(dataset->Cardinality(), GetParam().expected_cardinality);
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetCardinalityTest,
```

Comment:
Instead of defining different test classes for each of these types of tests (`DatasetCardinalityTest`, `IteratorOutputDtypesTest`, etc), which are all then instantiated with the same parameterized values, can you define one parameterized test class and write different tests for it?

e.g. 

```
class ParameterizedDatasetTest : ConcatenateDatasetOpTesthlper, 
                                      ::testing::WithParamInterface<TestParam>{}

TEST_P(ParameterizedDatasetTest, Cardinality) {
  ...
}

TEST_P(ParameterizedDatasetTest, IteratorOutputDtypes) {
  ...
}

TEST_P(ParameterizedDatasetTest, IteratorOutputShapes) {
  ...
}

INSTANTIATE_TEST_CASE_P(ParameterizedDatasetTest, ParameterizedDatasetTest,
                                   ::testing::ValuesIn(TestCases));

```